File: examples/canonical-auth/python/src/auth/claims.py
from dataclasses import dataclass
from datetime import datetime
from typing import Optional

@dataclass
class AuthContext:
    sub: str
    email: str
    tenant_id: Optional[str]
    tenant_role: str        # "super_admin" | "tenant_admin" | "tenant_member"
    groups: list[str]
    permissions: list[str]
    acl_ver: int
    issued_at: datetime
    expires_at: datetime

    def has_permission(self, permission: str) -> bool:
        return permission in self.permissions

File: examples/canonical-auth/rust/src/auth/claims.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Claims {
    pub iss: String,
    pub aud: Vec<String>,
    pub sub: String,
    pub exp: usize,
    pub iat: usize,
    pub nbf: usize,
    pub jti: String,
    pub tenant_id: Option<String>,
    pub tenant_role: String,
    pub groups: Vec<String>,
    pub permissions: Vec<String>,
    pub acl_ver: i32,
}

#[derive(Debug, Clone)]
pub struct AuthContext {
    pub sub: String,
    pub email: String,
    pub tenant_id: Option<String>,
    pub tenant_role: String,
    pub groups: Vec<String>,
    pub permissions: Vec<String>,
    pub acl_ver: i32,
    pub issued_at: DateTime<Utc>,
    pub expires_at: DateTime<Utc>,
}

impl AuthContext {
    pub fn has_permission(&self, permission: &str) -> bool {
        self.permissions.iter().any(|p| p == permission)
    }
}
